samples/snippets/sample_custom_types.py

```diff
@@ -0,0 +1,527 @@
+            )
+
+        # Memory must be a multiple of 256 MB
+        if self.memory_mb % 256 != 0:
+            raise RuntimeError("Requested memory must be a multiple of 256 MB.")
+
+        # Check if the requested memory isn't too little
+        if self.memory_mb < self.core_count * self.limits.min_mem_per_core:
+            raise RuntimeError(
+                f"Requested memory is too low. Minimal memory for {self.cpu_series.name} is {self.limits.min_mem_per_core} MB per core."
+            )
+
+        # Check if the requested memory isn't too much
+        if self.memory_mb > self.core_count * self.limits.max_mem_per_core:
+            if self.limits.allow_extra_memory:
+                if self.memory_mb > self.limits.extra_memory_limit:
+                    raise RuntimeError(
+                        f"Requested memory is too large.. Maximum memory allowed for {self.cpu_series.name} is {self.limits.extra_memory_limit} MB."
+                    )
+            else:
+                raise RuntimeError(
+                    f"Requested memory is too large.. Maximum memory allowed for {self.cpu_series.name} is {self.limits.max_mem_per_core} MB per core."
+                )
+
+    def __str__(self) -> str:
+        """
+        Return the custom machine type in form of a string acceptable by Compute Engine API.
+        """
+        if self.cpu_series in {
+            self.CPUSeries.E2_SMALL,
+            self.CPUSeries.E2_MICRO,
+            self.CPUSeries.E2_MEDIUM,
+        }:
+            return f"zones/{self.zone}/machineTypes/{self.cpu_series.value}-{self.memory_mb}"
+
+        if self.extra_memory_used:
+            return f"zones/{self.zone}/machineTypes/{self.cpu_series.value}-{self.core_count}-{self.memory_mb}-ext"
+
+        return f"zones/{self.zone}/machineTypes/{self.cpu_series.value}-{self.core_count}-{self.memory_mb}"
+
+    def short_type_str(self) -> str:
+        """
+        Return machine type in a format without the zone. For example, n2-custom-0-10240.
+        This format is used to create instance templates.
+        """
+        return str(self).rsplit("/", maxsplit=1)[1]
+
+    @classmethod
+    def from_str(cls, machine_type: str):
+        """
+        Construct a new object from a string. The string needs to be a valid custom machine type like:
+         - https://www.googleapis.com/compute/v1/projects/diregapic-mestiv/zones/us-central1-b/machineTypes/e2-custom-4-8192
+         - zones/us-central1-b/machineTypes/e2-custom-4-8192
+         - e2-custom-4-8192 (in this case, the zone parameter will not be set)
+        """
+        zone = None
+        if machine_type.startswith("http"):
+            machine_type = machine_type[machine_type.find("zones/") :]
+
+        if machine_type.startswith("zones/"):
+            _, zone, _, type_desc = machine_type.split("/")
+        else:
+            type_desc = machine_type
```

Comment:
do you mean maching_type = type_desc?

Reply:
I don't think I understand what you mean. I have refactored the code though, as I don't need to switch from using `machine_type` variable to `type_desc`. I can just keep using `machine_type`.